pkg/cloud/metadata.go

```
@@ -1,10 +1,176 @@
+	if !checkAzureMetadata() {
+		return "", ErrNotInCloud
+	}
+
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return "", err
+	}
+	client, err := armsubscriptions.NewClient(cred, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// List subscriptions and return the first one
+	pager := client.NewListPager(nil)
+	for pager.More() {
+		resp, err := pager.NextPage(context.TODO())
+		if err != nil {
+			return "", err
+		}
+		if len(resp.Value) > 0 {
+			return *resp.Value[0].SubscriptionID, nil
+		}
+	}
+	return "", fmt.Errorf("no Azure subscription found: %w", ErrNotInCloud)
+}
+
+// GetGCPProjectID retrieves the GCP numerical project ID.
+func GetGCPProjectID() (string, error) {
+	if !metadata.OnGCE() {
+		return "", ErrNotInCloud
+	}
+	return metadata.NumericProjectID()
+}
+
+// checkAzureMetadata detects Azure by querying IMDS.
+func checkAzureMetadata() bool {
+	client := http.Client{Timeout: 1 * time.Second}
+	req, err := http.NewRequest(http.MethodGet, "http://169.254.169.254/metadata/instance?api-version=2021-02-01", nil)
+	if err != nil {
+		return false
+	}
+	req.Header.Set("Metadata", "true")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return false
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	return resp.StatusCode == http.StatusOK
+}
+
+// init registers the built-in cloud detectors
+// maintained the original order: GCP first, then AWS, then Azure
+func init() {
+	RegisterDetector(GCPCloud, GetGCPProjectID)
+	RegisterDetector(AWSCloud, GetAWSAccountID)
+	RegisterDetector(AzureCloud, GetAzureSubscriptionID)
+}
+
+// Detect runs the cloud detection logic and caches the result.
+// Detectors are tried in registration order
+func Detect() {
+	// Iterate through detectors in the order they were registered
+	for _, name := range detectorOrder {
+		detector := detectorsRegistry[name]
+		if id, err := detector(); err == nil {
+			cloudType, cloudID, cloudDetected = name, id, true
+			return
+		}
```

If we're ignoring the err on purpose, maybe it's worth documenting.